core/src/nucleus/actions/validate.rs

```diff
@@ -3,122 +3,162 @@ use crate::{
+            )));
         }
+    }
 
-        EntryType::Deletion => {
-            // FIXME
-        }
+    Ok(())
+}
 
-        EntryType::CapTokenGrant => {
-            // FIXME
-        }
+fn spawn_validation_ribosome(
+    id: ProcessUniqueId,
+    entry: Entry,
+    validation_data: ValidationData,
+    context: Arc<Context>,
+) {
+    thread::spawn(move || {
+        let address = entry.address();
+        let maybe_validation_result = callback::validate_entry::validate_entry(
+            entry.clone(),
+            validation_data.clone(),
+            context.clone(),
+        );
 
-        EntryType::AgentId => {
-            // FIXME
-        }
-        _ => {
-            return FutureObj::new(Box::new(future::err(HolochainError::ValidationFailed(
-                format!(
-                    "Attempted to validate system entry type {:?}",
-                    entry.entry_type(),
-                ),
-            ))));
-        }
-    }
+        let result = match maybe_validation_result {
+            Ok(validation_result) => match validation_result {
+                CallbackResult::Fail(error_string) => Err(error_string),
+                CallbackResult::Pass => Ok(()),
+                CallbackResult::NotImplemented(reason) => Err(format!(
+                    "Validation callback not implemented for {:?} ({})",
+                    entry.entry_type().clone(),
+                    reason
+                )),
+                _ => unreachable!(),
+            },
+            Err(error) => Err(error.to_string()),
+        };
+
+        context
+            .action_channel()
+            .send(ActionWrapper::new(Action::ReturnValidationResult((
+                (id, address),
+                result,
+            ))))
+            .expect("action channel to be open in reducer");
+    });
+}
+
+fn validate_provenances(validation_data: &ValidationData) -> Result<(), HolochainError> {
+    let header = &validation_data.package.chain_header;
+    header
+        .provenances()
+        .iter()
+        .map(|provenance| {
+            let author = &provenance.0;
+            let signature = &provenance.1;
+            let signature_string: String = signature.clone().into();
+            let signature_bytes: Vec<u8> = base64::decode(&signature_string).map_err(|_| {
+                HolochainError::ValidationFailed("Signature syntactically invalid".to_string())
+            })?;
+
+            let mut signature_buf = SecBuf::with_insecure(signature_bytes.len());
+            signature_buf
+                .write(0, signature_bytes.as_slice())
+                .expect("SecBuf must be writeable");
+
+            let mut message_buf =
+                SecBuf::with_insecure_from_string(header.entry_address().to_string());
+            let result = Keypair::verify(author.to_string(), &mut signature_buf, &mut message_buf)?;
+
+            (result == 0).ok_or(HolochainError::ValidationFailed(
+                "Signature invalid".to_string(),
```

Comment:
Can we add something here about which of the provenences is invalid?  Like:
```suggestion
                format!("Signature for {} invalid",author),
```